tls/s2n_ktls.c

```diff
@@ -15,11 +15,176 @@
+
+            /* The input stuffer should be empty before enabling kTLS. */
+            RESULT_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        /* retrieve the receive fd */
+        RESULT_ENSURE_REF(conn->recv_io_context);
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        /* retrieve the send fd */
+        RESULT_ENSURE_REF(conn->send_io_context);
+        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
+        *fd = peer_socket_ctx->fd;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* If already enabled then return success */
+    if (ktls_mode == S2N_KTLS_MODE_SEND && conn->ktls_send_enabled) {
+        RESULT_BAIL(S2N_ERR_KTLS_ALREADY_ENABLED);
+    }
+    if (ktls_mode == S2N_KTLS_MODE_RECV && conn->ktls_recv_enabled) {
+        RESULT_BAIL(S2N_ERR_KTLS_ALREADY_ENABLED);
+    }
+
+    int fd = 0;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+
+    /* Calls to setsockopt require a real socket, which is not used in unit tests. */
+    RESULT_ENSURE(!disable_ktls_socket_config_for_testing, S2N_ERR_KTLS_DISABLED_FOR_TEST);
+
+    /* Enable 'tls' ULP for the socket. https://lwn.net/Articles/730207 */
+    int ret = setsockopt(fd, S2N_SOL_TCP, S2N_TCP_ULP, S2N_TLS_ULP_NAME, S2N_TLS_ULP_NAME_SIZE);
+
+    if (ret != 0) {
+        /* EEXIST: https://man7.org/linux/man-pages/man3/errno.3.html
+         *
+         * TCP_ULP has already been enabled on the socket so the operation is a noop.
+         * Since its possible to call this twice, once for TX and once for RX, consider
+         * the noop a success. */
+        if (errno != EEXIST) {
+            RESULT_BAIL(S2N_ERR_KTLS_ULP);
+        }
+    }
+
+    return S2N_RESULT_OK;
+}
+
+/*
+ * Since kTLS is an optimization, it is possible to continue operation
+ * by using userspace TLS if kTLS is not supported. Upon successfully
+ * enabling kTLS, we set connection->ktls_send_enabled (and recv) to true.
+ *
+ * kTLS configuration errors are recoverable since calls to setsockopt are
+ * non-destructive and its possible to fallback to userspace.
+ */
+int s2n_connection_ktls_enable_send(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+
+    if (!s2n_ktls_is_supported_on_platform()) {
+        POSIX_BAIL(S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+    }
+
+    POSIX_GUARD_RESULT(s2n_ktls_validate(conn, S2N_KTLS_MODE_SEND));
+
```

Comment:
```suggestion
```
General nit, is it just me or is there a bit of extra whitespace in this PR? I would take a pass through and cut down on the extra space if you can.

Reply:
its clang formatted so it should be correctly formatted. However I can remove the empty line.